src/python/lib/ensembl/compara/hal/maf_to_fasta.py

```diff
@@ -0,0 +1,247 @@
+    a dot in either the genome or sequence name. Taking the names of genomes known to be in a MAF file,
+    this function creates a string Pattern object that can be used to extract the genome and sequence
+    names from a MAF src field of the form '<genome>.<seqid>'.
+
+    Args:
+        genome_names: The genome names expected to be in the input MAF file.
+
+    Returns:
+        Compiled string Pattern object.
+
+    Raises:
+        ValueError: If any of the input genome names is a prefix of any other genome name
+            and that prefix is followed by a dot ('.') in the longer genome name, the longer
+            genome name can never match unambiguously, and this function cannot be used to
+            extract the genome and sequence names. For example, genomes 'genomeA' and
+            'genomeA.1' would both match src field 'genomeA.1.1'.
+
+    Example::
+        >>> maf_src_regex = compile_maf_src_regex(['genomeA.1', 'genomeB.1'])
+        >>> match = maf_src_regex.match('genomeA.1.1')
+        >>> match['genome']
+        'genomeA.1'
+        >>> match['seqid']
+        '1'
+
+    """
+    for genome_name in genome_names:
+        dot_regex = re.compile('[.]')
+        for match in dot_regex.finditer(genome_name):
+            prefix = genome_name[:match.start()]
+            if prefix in genome_names:
+                raise ValueError(f"cannot create a MAF src regex - genome name '{prefix}'"
+                                 f" is a prefix of '{genome_name}'")
+    genome_patt = '|'.join(map(re.escape, genome_names))
+    if not genome_patt:
+        raise ValueError("'genome_names' must be an iterable containing at least one genome name")
+    maf_src_patt = f'^(?P<genome>{genome_patt})[.](?P<seqid>.+)$'
+    return re.compile(maf_src_patt)
+
+
+def convert_maf_to_fasta(maf_file: PathLike, output_dir: PathLike,
+                         genomes_file: Optional[PathLike] = None) -> None:
+    """Convert each block of a MAF alignment to a FASTA file.
+
+    Each block is stored in a FASTA file (e.g. '123.fa'), with its metadata stored in a
+    matching JSON file (e.g. '123.json'), where both file names have the same file stem.
+
+    Args:
+        maf_file: Input MAF file with alignment blocks. The src fields of
+            this MAF file should be of the form '<genome>.<seqid>'.
+        output_dir: Output directory under which FASTA and JSON files will be created.
+        genomes_file: File listing the genomes in the input MAF file, one per line. This is used
+            to compile a regex that splits MAF src fields of the form '<genome>.<seqid>'
+            into their component parts. If any of the genomes or their DNA assembly
+            sequences in the MAF file contains a dot ('.'), this is required.
+
+    Raises:
+        ValueError: If a MAF src field cannot be parsed.
+
+    """
+
+    if genomes_file is not None:
```

Comment:
if `genome_file` is `None`?

Reply:
If `genomes_file` is `None`, and the input MAF file contains src fields with only one dot (`.`) (e.g. `genomeA.chr1`), then the src fields are split into two either side of the dot, and the substring before the dot is taken as the genome name, while the substring after the dot is taken as the DNA assembly sequence name.

If `genomes_file` is `None` and the input MAF file contains src fields with multiple dots (e.g. `genomeA.1.chr1`), it is impossible to split the src field into two parts either side of a dot (which dot to use?), so a `ValueError` is raised: `MAF src field parse failed due to multiple dot separators...`.